## 1. The NIST Special Publication 800-63B
<p>If you – 50 years ago – needed to come up with a secret password you were probably part of a secret espionage organization or (more likely) you were pretending to be a spy when playing as a kid. Today, many of us are forced to come up with new passwords <em>all the time</em> when signing into sites and apps. As a password <em>inventeur</em> it is your responsibility to come up with good, hard-to-crack passwords. But it is also in the interest of sites and apps to make sure that you use good passwords. The problem is that it's really hard to define what makes a good password. However, <em>the National Institute of Standards and Technology</em> (NIST) knows what the second best thing is: To make sure you're at least not using a <em>bad</em> password. </p>
<p>In this notebook, we will go through the rules in <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a> which details what checks a <em>verifier</em> (what the NIST calls a second party responsible for storing and verifying passwords) should perform to make sure users don't pick bad passwords. We will go through the passwords of users from a fictional company and use R to flag the users with bad passwords. But us being able to do this already means the fictional company is breaking one of the rules of 800-63B:</p>
<blockquote>
  <p>Verifiers SHALL store memorized secrets in a form that is resistant to offline attacks. Memorized secrets SHALL be salted and hashed using a suitable one-way key derivation function.</p>
</blockquote>
<p>That is, never save users' passwords in plaintext, always encrypt the passwords! Keeping this in mind for the next time we're building a password management system, let's load in the data.</p>
<p><em>Warning: The list of passwords and the fictional user database both contain <strong>real</strong> passwords leaked from <strong>real</strong> websites. These passwords have not been filtered in any way and include words that are explicit, derogatory and offensive.</em></p>

In [1]:
# Importing the tidyverse library

library(tidyverse)

# Loading in datasets/users.csv 

users <- read_csv("datasets/users.csv")

# Counting how many users we've got

nrow(users)

# Taking a look at the 12 first users

head(users, 12)

── Attaching packages ─────────────────────────────────────── tidyverse 1.2.1 ──

✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.0.6     ✔ dplyr   1.0.4
✔ tidyr   1.1.2     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


── Column specification ────────────────────────────────────────────────────────
cols(
  id = col_double(),
  user_name = col_character(),
  password = col_character()
)




[1] 982

id user_name        password        
1   1 vance.jennings   joobheco        
2   2 consuelo.eaton   0869347314      
3   3 mitchel.perkins  fabypotter      
4   4 odessa.vaughan   aharney88       
5   5 araceli.wilder   acecdn3000      
6   6 shawn.harrington 5278049         
7   7 evelyn.gay       master          
8   8 noreen.hale      murphy          
9   9 gladys.ward      lwsves2         
10 10 brant.zimmerman  1190KAREN5572497
11 11 leanna.abbott    aivlys24        
12 12 milford.hubbard  hubbard

## 2. Passwords should not be too short
<p>If we take a look at the first 12 users above we already see some bad passwords. But let's not get ahead of ourselves and start flagging passwords <em>manually</em>. What is the first thing we should check according to the NIST Special Publication 800-63B?</p>
<blockquote>
  <p>Verifiers SHALL require subscriber-chosen memorized secrets to be at least 8 characters in length.</p>
</blockquote>
<p>Ok, so the passwords of our users shouldn't be too short. Let's start by checking that!</p>

In [2]:
# Calculating the lengths of users' passwords
# Flagging the users with too short passwords

users <- users %>%
    mutate(
        length = str_length(password),
        too_short = case_when(
            length < 8 ~ TRUE,
            TRUE ~ FALSE
        )
    )

# Counting the number of users with too short passwords

users %>%
    count(too_short)

# Taking a look at the 12 first rows

head(users, 12)

too_short n  
1 FALSE     606
2  TRUE     376

id user_name        password         length too_short
1   1 vance.jennings   joobheco          8     FALSE    
2   2 consuelo.eaton   0869347314       10     FALSE    
3   3 mitchel.perkins  fabypotter       10     FALSE    
4   4 odessa.vaughan   aharney88         9     FALSE    
5   5 araceli.wilder   acecdn3000       10     FALSE    
6   6 shawn.harrington 5278049           7      TRUE    
7   7 evelyn.gay       master            6      TRUE    
8   8 noreen.hale      murphy            6      TRUE    
9   9 gladys.ward      lwsves2           7      TRUE    
10 10 brant.zimmerman  1190KAREN5572497 16     FALSE    
11 11 leanna.abbott    aivlys24          8     FALSE    
12 12 milford.hubbard  hubbard           7      TRUE

## 3.  Common passwords people use
<p>Already this simple rule flagged a couple of offenders among the first 12 users. Next up in Special Publication 800-63B is the rule that</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets against a list that contains values known to be commonly-used, expected, or compromised.</p>
  <ul>
  <li>Passwords obtained from previous breach corpuses.</li>
  <li>Dictionary words.</li>
  <li>Repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</li>
  <li>Context-specific words, such as the name of the service, the username, and derivatives thereof.</li>
  </ul>
</blockquote>
<p>We're going to check these in order and start with <em>Passwords obtained from previous breach corpuses</em>, that is, websites where hackers have leaked all the users' passwords. As many websites don't follow the NIST guidelines and encrypt passwords there now exist large lists of the most popular passwords. Let's start by loading in the 10,000 most common passwords which I've taken from <a href="https://github.com/danielmiessler/SecLists/tree/master/Passwords">here</a>.</p>

In [3]:
# Reading in the top 10000 passwords

common_passwords <-  read_lines("datasets/10_million_password_list_top_10000.txt")

# Taking a look at the top 100

head(common_passwords, 100)

[1] "123456"     "password"   "12345678"   "qwerty"     "123456789" 
  [6] "12345"      "1234"       "111111"     "1234567"    "dragon"    
 [11] "123123"     "baseball"   "abc123"     "football"   "monkey"    
 [16] "letmein"    "696969"     "shadow"     "master"     "666666"    
 [21] "qwertyuiop" "123321"     "mustang"    "1234567890" "michael"   
 [26] "654321"     "pussy"      "superman"   "1qaz2wsx"   "7777777"   
 [31] "fuckyou"    "121212"     "000000"     "qazwsx"     "123qwe"    
 [36] "killer"     "trustno1"   "jordan"     "jennifer"   "zxcvbnm"   
 [41] "asdfgh"     "hunter"     "buster"     "soccer"     "harley"    
 [46] "batman"     "andrew"     "tigger"     "sunshine"   "iloveyou"  
 [51] "fuckme"     "2000"       "charlie"    "robert"     "thomas"    
 [56] "hockey"     "ranger"     "daniel"     "starwars"   "klaster"   
 [61] "112233"     "george"     "asshole"    "computer"   "michelle"  
 [66] "jessica"    "pepper"     "1111"       "zxcvbn"     "555555"    
 [71] "11111111"   "131313"     "freedom"    "777777"     "pass"      
 [76] "fuck"       "maggie"     "159753"     "aaaaaa"     "ginger"    
 [81] "princess"   "joshua"     "cheese"     "amanda"     "summer"    
 [86] "love"       "ashley"     "6969"       "nicole"     "chelsea"   
 [91] "biteme"     "matthew"    "access"     "yankees"    "987654321" 
 [96] "dallas"     "austin"     "thunder"    "taylor"     "matrix"

## 4.  Passwords should not be common passwords
<p>The list of passwords was ordered, with the most common passwords first, and so we shouldn't be surprised to see passwords like <code>123456</code> and <code>qwerty</code> above. As hackers also have access to this list of common passwords, it's important that none of our users use these passwords!</p>
<p>Let's flag all the passwords in our user database that are among the top 10,000 used passwords.</p>

In [4]:
# Flagging the users with passwords that are common passwords
#users$common_password <- ifelse(users$password %in% common_passwords, TRUE, FALSE)

users <- users %>%
    mutate(
        common_password = case_when(
            password %in% common_passwords ~ TRUE,
            TRUE ~ FALSE
        )
    )
# Counting the number of users using common passwords

users %>%
    count(common_password)

# Taking a look at the 12 first rows

head(users, 12)

common_password n  
1 FALSE           853
2  TRUE           129

id user_name        password         length too_short common_password
1   1 vance.jennings   joobheco          8     FALSE     FALSE          
2   2 consuelo.eaton   0869347314       10     FALSE     FALSE          
3   3 mitchel.perkins  fabypotter       10     FALSE     FALSE          
4   4 odessa.vaughan   aharney88         9     FALSE     FALSE          
5   5 araceli.wilder   acecdn3000       10     FALSE     FALSE          
6   6 shawn.harrington 5278049           7      TRUE     FALSE          
7   7 evelyn.gay       master            6      TRUE      TRUE          
8   8 noreen.hale      murphy            6      TRUE      TRUE          
9   9 gladys.ward      lwsves2           7      TRUE     FALSE          
10 10 brant.zimmerman  1190KAREN5572497 16     FALSE     FALSE          
11 11 leanna.abbott    aivlys24          8     FALSE     FALSE          
12 12 milford.hubbard  hubbard           7      TRUE     FALSE

## 5. Passwords should not be common words
<p>Ay ay ay! It turns out many of our users use common passwords, and of the first 12 users there are already two. However, as most common passwords also tend to be short, they were already flagged as being too short. What is the next thing we should check?</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains […] dictionary words.</p>
</blockquote>
<p>This follows the same logic as before: It is easy for hackers to check users' passwords against common English words and therefore common English words make bad passwords. Let's check our users' passwords against the top 10,000 English words from <a href="https://github.com/first20hours/google-10000-english">Google's Trillion Word Corpus</a>.</p>

In [5]:
# Reading in a list of the 10000 most common words

words <- read_lines("datasets/google-10000-english.txt")

# Flagging the users with passwords that are common words

users <- users %>%
    mutate(
        common_word = case_when(
            str_to_lower(password) %in% words ~ TRUE,
            TRUE ~ FALSE
        )
    )

# Counting the number of users using common words as passwords

users %>%
    count(common_word)

# Taking a look at the 12 first rows

head(users, 12)

common_word n  
1 FALSE       845
2  TRUE       137

id user_name        password         length too_short common_password
1   1 vance.jennings   joobheco          8     FALSE     FALSE          
2   2 consuelo.eaton   0869347314       10     FALSE     FALSE          
3   3 mitchel.perkins  fabypotter       10     FALSE     FALSE          
4   4 odessa.vaughan   aharney88         9     FALSE     FALSE          
5   5 araceli.wilder   acecdn3000       10     FALSE     FALSE          
6   6 shawn.harrington 5278049           7      TRUE     FALSE          
7   7 evelyn.gay       master            6      TRUE      TRUE          
8   8 noreen.hale      murphy            6      TRUE      TRUE          
9   9 gladys.ward      lwsves2           7      TRUE     FALSE          
10 10 brant.zimmerman  1190KAREN5572497 16     FALSE     FALSE          
11 11 leanna.abbott    aivlys24          8     FALSE     FALSE          
12 12 milford.hubbard  hubbard           7      TRUE     FALSE          
   common_word
1  FALSE      
2  FALSE      
3  FALSE      
4  FALSE      
5  FALSE      
6  FALSE      
7   TRUE      
8   TRUE      
9  FALSE      
10 FALSE      
11 FALSE      
12 FALSE

## 6. Passwords should not be your name
<p>It turns out many of our passwords were common English words too! Next up on the NIST list:</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains […] context-specific words, such as the name of the service, the username, and derivatives thereof.</p>
</blockquote>
<p>Ok, so there are many things we could check here. One thing to notice is that our users' usernames consist of their first names and last names separated by a dot. For now, let's just flag passwords that are the same as either a user's first or last name.</p>

In [6]:
# Extracting first and last names into their own columns

names <- as.data.frame(str_split(users$user_name, "\\.",  simplify = TRUE))
users$first_name <- names$V1
users$last_name <- names$V2

# Flagging the users with passwords that matches their names

users <- users %>%
    mutate(
        uses_name = case_when(
            str_to_lower(password) == first_name ~ TRUE,
            str_to_lower(password) == last_name ~ TRUE,
            TRUE ~ FALSE
        )
    )

# Counting the number of users using names as passwords

users %>%
    count(uses_name)

# Taking a look at the 12 first rows

head(users, 12)

uses_name n  
1 FALSE     932
2  TRUE      50

id user_name        password         length too_short common_password
1   1 vance.jennings   joobheco          8     FALSE     FALSE          
2   2 consuelo.eaton   0869347314       10     FALSE     FALSE          
3   3 mitchel.perkins  fabypotter       10     FALSE     FALSE          
4   4 odessa.vaughan   aharney88         9     FALSE     FALSE          
5   5 araceli.wilder   acecdn3000       10     FALSE     FALSE          
6   6 shawn.harrington 5278049           7      TRUE     FALSE          
7   7 evelyn.gay       master            6      TRUE      TRUE          
8   8 noreen.hale      murphy            6      TRUE      TRUE          
9   9 gladys.ward      lwsves2           7      TRUE     FALSE          
10 10 brant.zimmerman  1190KAREN5572497 16     FALSE     FALSE          
11 11 leanna.abbott    aivlys24          8     FALSE     FALSE          
12 12 milford.hubbard  hubbard           7      TRUE     FALSE          
   common_word first_name last_name  uses_name
1  FALSE       vance      jennings   FALSE    
2  FALSE       consuelo   eaton      FALSE    
3  FALSE       mitchel    perkins    FALSE    
4  FALSE       odessa     vaughan    FALSE    
5  FALSE       araceli    wilder     FALSE    
6  FALSE       shawn      harrington FALSE    
7   TRUE       evelyn     gay        FALSE    
8   TRUE       noreen     hale       FALSE    
9  FALSE       gladys     ward       FALSE    
10 FALSE       brant      zimmerman  FALSE    
11 FALSE       leanna     abbott     FALSE    
12 FALSE       milford    hubbard     TRUE

## 7. Passwords should not be repetitive
<p>Milford Hubbard (user number 12 above), what where you thinking!? Ok, so the last thing we are going to check is a bit tricky:</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets [so that they don't contain] repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</p>
</blockquote>
<p>This is tricky to check because what is <em>repetitive</em> is hard to define. Is <code>11111</code> repetitive? Yes! Is <code>12345</code> repetitive? Well, kind of. Is <code>13579</code> repetitive? Maybe not..? To check for <em>repetitiveness</em> can be arbitrarily complex, but here we're only going to do something simple. We're going to flag all passwords that contain 4 or more repeated characters.</p>

In [24]:
# Splitting the passwords into vectors of single characters
split_passwords <- str_extract_all(users$password, boundary("character"))

# Picking out the max number of repeat characters for each password


users$max_repeats <- sapply(split_passwords, function(split_password) {
    return(max(rle(split_password)$lengths))
})


# Flagging the passwords with >= 4 repeats
 
users <- users %>%
    mutate(
        too_many_repeats = case_when(
            max_repeats >= 4 ~ TRUE,
            TRUE ~ FALSE
        )
    )


# Taking a look at the users with too many repeats
users %>%
    filter(too_many_repeats == TRUE)

id  user_name        password length too_short common_password common_word
1 147 patti.dixon      555555   6       TRUE      TRUE           FALSE      
2 573 cornelia.bradley 555555   6       TRUE      TRUE           FALSE      
3 645 essie.lopez      11111    5       TRUE      TRUE           FALSE      
4 799 charley.key      888888   6       TRUE      TRUE           FALSE      
5 808 thurman.osborne  rinnnng0 8      FALSE     FALSE           FALSE      
6 942 mitch.ferguson   aaaaaa   6       TRUE      TRUE           FALSE      
  first_name last_name uses_name max_repeats too_many_repeats
1 patti      dixon     FALSE     6           TRUE            
2 cornelia   bradley   FALSE     6           TRUE            
3 essie      lopez     FALSE     5           TRUE            
4 charley    key       FALSE     6           TRUE            
5 thurman    osborne   FALSE     4           TRUE            
6 mitch      ferguson  FALSE     6           TRUE

## 8. All together now!
<p>Now we have implemented all the basic tests for bad passwords suggested by NIST Special Publication 800-63B! What's left is just to flag all bad passwords and maybe to send these users an e-mail that strongly suggests they change their password.</p>

In [28]:
# Flagging all passwords that are bad
#users$bad_password <- ....
users <- users %>%
    mutate(
        bad_password = case_when(
            (too_short + common_password + common_word + uses_name + too_many_repeats) == 0 ~ FALSE,
            TRUE ~ TRUE
        )
    )
# Counting the number of bad passwords
users %>%
    count(bad_password)

# Looking at the first 100 bad passwords
users %>%
    filter(bad_password == TRUE) %>%
    slice(1:100)

bad_password n  
1 FALSE        558
2  TRUE        424

id  user_name           password     length too_short common_password
1    6  shawn.harrington    5278049      7       TRUE     FALSE          
2    7  evelyn.gay          master       6       TRUE      TRUE          
3    8  noreen.hale         murphy       6       TRUE      TRUE          
4    9  gladys.ward         lwsves2      7       TRUE     FALSE          
5   12  milford.hubbard     hubbard      7       TRUE     FALSE          
6   14  jamie.cochran       310356       6       TRUE     FALSE          
7   16  lorrie.gay          oZ4k0QE      7       TRUE     FALSE          
8   17  domingo.dyer        chelsea      7       TRUE      TRUE          
9   18  martin.pacheco      zvc1939      7       TRUE     FALSE          
10  19  shelby.massey       nickgd       6       TRUE     FALSE          
11  22  leticia.sanford     cocacola     8      FALSE      TRUE          
12  23  jenny.woodard       woodard      7       TRUE     FALSE          
13  26  dianna.munoz        AJ9Da        5       TRUE     FALSE          
14  27  julia.savage        ewokzs       6       TRUE     FALSE          
15  29  joaquin.walters     YyGjz8E      7       TRUE     FALSE          
16  31  rosanna.reid        reid         4       TRUE     FALSE          
17  35  roger.golden        jOYZBs8      7       TRUE     FALSE          
18  39  gus.padilla         wwewwf1      7       TRUE     FALSE          
19  44  vito.nicholson      225377       6       TRUE     FALSE          
20  46  morris.price        NdZ7E6       6       TRUE     FALSE          
21  48  jarred.white        CQB3Z        5       TRUE     FALSE          
22  49  junior.burch        diffo        5       TRUE     FALSE          
23  52  ahmad.hopper        123456789    9      FALSE      TRUE          
24  53  rosario.merrill     y8uM7D6      7       TRUE     FALSE          
25  57  dwayne.hurst        mikeloo      7       TRUE     FALSE          
26  59  rosalinda.rodriquez golden       6       TRUE      TRUE          
27  60  landon.middleton    apr2502      7       TRUE     FALSE          
28  63  dominique.dawson    jfasii       6       TRUE     FALSE          
29  64  reuben.higgins      sanaang      7       TRUE     FALSE          
30  65  ursula.wood         jecarn       6       TRUE     FALSE          
⋮   ⋮   ⋮                   ⋮            ⋮      ⋮         ⋮              
71  164 dixie.noble         dixie         5      TRUE      TRUE          
72  165 erick.miles         gtapbt        6      TRUE     FALSE          
73  166 major.daniels       123654        6      TRUE      TRUE          
74  171 ora.stuart          1036900       7      TRUE     FALSE          
75  173 tanya.gaines        flavor        6      TRUE     FALSE          
76  176 sharron.bryant      FGPb9gX       7      TRUE     FALSE          
77  179 sammie.montgomery   vHO2y         5      TRUE     FALSE          
78  182 loretta.dickson     precise       7      TRUE     FALSE          
79  183 thanh.mason         190664        6      TRUE     FALSE          
80  184 pierre.yates        4f3aRNG       7      TRUE     FALSE          
81  185 charlotte.morin     austin        6      TRUE      TRUE          
82  187 carmella.cline      jtc123        6      TRUE     FALSE          
83  188 reuben.wells        eLvtn         5      TRUE     FALSE          
84  190 reyna.holloway      122412        6      TRUE     FALSE          
85  191 marlene.faulkner    8309113       7      TRUE     FALSE          
86  194 anderson.dillard    dillard       7      TRUE     FALSE          
87  197 jamie.goodman       william       7      TRUE      TRUE          
88  198 tommie.ferrell      NCAA50        6      TRUE     FALSE          
89  201 brenda.blevins      4101516       7      TRUE     FALSE          
90  204 lane.howard         leyvi         5      TRUE     FALSE          
91  205 denis.stone         ancan         5      TRUE     FALSE          
92  206 blake.donaldson     blake         5      TRUE      TRUE          
93  209 

## 9. Otherwise, the password should be up to the user
<p>In this notebook, we've implemented the password checks recommended by the NIST Special Publication 800-63B. It's certainly possible to better implement these checks, for example, by using a longer list of common passwords. Also note that the NIST checks in no way guarantee that a chosen password is good, just that it's not obviously bad.</p>
<p>Apart from the checks we've implemented above the NIST is also clear with what password rules should <em>not</em> be imposed:</p>
<blockquote>
  <p>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets. Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically).</p>
</blockquote>
<p>So the next time a website or app tells you to "include both a number, symbol and an upper and lower case character in your password" you should send them a copy of <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a>.</p>

In [ ]:
# Enter a password that passes the NIST requirements
# PLEASE DO NOT USE AN EXISTING PASSWORD HERE
new_password <- "qwerty123"